// @iobroker/create-adapter/build/src/lib/packageVersions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasVersionSpecifier = hasVersionSpecifier;
exports.getVersionSpecifier = getVersionSpecifier;
exports.getPackageName = getPackageName;
exports.fetchPackageVersion = fetchPackageVersion;
exports.fetchPackageReferenceVersion = fetchPackageReferenceVersion;
const axios_1 = __importDefault(require("axios"));
const semver = __importStar(require("semver"));
const tools_1 = require("./tools");
const allVersionsCache = new Map();
const latestVersionCache = new Map();
/**
 *
 * @param packageName
 */
function hasVersionSpecifier(packageName) {
    return packageName.lastIndexOf("@") > 0;
}
/**
 *
 * @param packageNameAndVersion
 */
function getVersionSpecifier(packageNameAndVersion) {
    const atIndex = packageNameAndVersion.lastIndexOf("@");
    if (atIndex > 0) {
        return packageNameAndVersion.slice(atIndex + 1);
    }
}
/**
 * Returns only the package name from a string of the form [@scope/]package-name@version
 *
 * @param packageNameAndVersion
 */
function getPackageName(packageNameAndVersion) {
    const atIndex = packageNameAndVersion.lastIndexOf("@");
    if (atIndex > 0) {
        return packageNameAndVersion.slice(0, atIndex);
    }
    return packageNameAndVersion;
}
/**
 * Returns the latest version of an npm package
 *
 * @param packageName The npm package name
 * @param fallbackVersion The fallback version to return in case anything goes wrong. If this is set, no error is thrown.
 */
async function fetchAllPackageVersions(packageName) {
    if (allVersionsCache.has(packageName)) {
        return allVersionsCache.get(packageName);
    }
    const packageURIComponent = encodeURIComponent(packageName);
    const url = `https://registry.npmjs.org/${packageURIComponent}`;
    let options = { url, timeout: (0, tools_1.getRequestTimeout)() };
    // If an https-proxy is defined as an env variable, use it
    options = (0, tools_1.applyHttpsProxy)(options);
    const response = await axios_1.default.request(options);
    if (response.status !== 200) {
        throw new Error(`Failed to fetch the versions for ${packageName} (${response.status})`);
    }
    const allVersions = Object.keys(response.data.versions);
    allVersionsCache.set(packageName, allVersions);
    return allVersions;
}
async function fetchSpecificPackageVersion(packageNameAndVersion, fallbackVersion) {
    // A specific version is requested, return the highest version matching the specifier
    try {
        const versionSpecifier = getVersionSpecifier(packageNameAndVersion);
        const packageName = getPackageName(packageNameAndVersion);
        const allVersions = await fetchAllPackageVersions(packageName);
        const ret = semver.maxSatisfying(allVersions, versionSpecifier);
        if (!ret) {
            throw new Error(`No version of ${packageName} found matching ${versionSpecifier}`);
        }
        return ret;
    }
    catch (e) {
        if (fallbackVersion) {
            return fallbackVersion;
        }
        throw e;
    }
}
async function fetchLatestPackageVersion(packageName, fallbackVersion) {
    if (latestVersionCache.has(packageName)) {
        return latestVersionCache.get(packageName);
    }
    const packageURIComponent = encodeURIComponent(packageName);
    const url = `https://registry.npmjs.org/-/package/${packageURIComponent}/dist-tags`;
    let options = { url, timeout: (0, tools_1.getRequestTimeout)() };
    // If an https-proxy is defined as an env variable, use it
    options = (0, tools_1.applyHttpsProxy)(options);
    const response = await axios_1.default.request(options);
    if (response.status !== 200) {
        if (fallbackVersion) {
            return fallbackVersion;
        }
        throw new Error(`Failed to fetch the version for ${packageName} (${response.status})`);
    }
    const version = response.data.latest;
    latestVersionCache.set(packageName, version);
    return version;
}
/**
 * Returns the latest version of an npm package
 *
 * @param packageName The npm package name
 * @param fallbackVersion The fallback version to return in case anything goes wrong. If this is set, no error is thrown.
 */
async function fetchPackageVersion(packageName, fallbackVersion) {
    if (hasVersionSpecifier(packageName)) {
        return fetchSpecificPackageVersion(packageName, fallbackVersion);
    }
    return fetchLatestPackageVersion(packageName, fallbackVersion);
}
/**
 * Returns the reference to the version of an npm package
 *
 * @param packageName The npm package name (may include a version specifier like `abc@1.2.3`)
 */
async function fetchPackageReferenceVersion(packageName) {
    if (hasVersionSpecifier(packageName)) {
        const versionSpecifier = getVersionSpecifier(packageName);
        const pureVersionSpec = versionSpecifier.replace(/^[\^~]/, "").trim();
        const versionDepth = pureVersionSpec.replace(/^(0\.)+/, "").split(".").length;
        const version = await fetchSpecificPackageVersion(packageName, pureVersionSpec);
        if (versionSpecifier.charAt(0) === "^" || versionDepth <= 1) {
            return `^${version}`;
        }
        else if (versionSpecifier.charAt(0) === "~" || versionDepth == 2) {
            return `~${version}`;
        }
        return version;
    }
    return `^${await fetchLatestPackageVersion(packageName, "0.0.0")}`;
}
