// @iobroker/adapter-dev/build/translators/google-v3-translator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleV3Translator = void 0;
const translate_1 = require("@google-cloud/translate");
const fs_extra_1 = require("fs-extra");
const translate_2 = require("../translate");
/**
 * @see Translator implementation that uses the Google Translation API.
 * This API requires credentials which must be stored in a file pointed to
 * by the environment variable GOOGLE_APPLICATION_CREDENTIALS.
 */
class GoogleV3Translator {
    async init() {
        this.credentials = await (0, fs_extra_1.readJson)(process.env.GOOGLE_APPLICATION_CREDENTIALS || "");
        this.translationClient = new translate_1.TranslationServiceClient();
    }
    async translate(text, targetLang) {
        var _a;
        try {
            const request = {
                parent: `projects/${this.credentials.project_id}/locations/global`,
                contents: [text],
                mimeType: "text/plain",
                sourceLanguageCode: "en",
                targetLanguageCode: targetLang,
            };
            const [response] = await this.translationClient.translateText(request);
            if (response.translations &&
                ((_a = response.translations[0]) === null || _a === void 0 ? void 0 : _a.translatedText)) {
                return response.translations[0].translatedText;
            }
            throw new Error(`Google couldn't translate "${text}"`);
        }
        catch (err) {
            // Handle Google API errors
            if (err.code === 8 || err.code === 429) {
                // Resource exhausted (quota exceeded) or rate limited
                throw new translate_2.RateLimitedError(`Google Translate quota exceeded or rate limited: ${err.message}`);
            }
            else if (err.code === 4) {
                // Deadline exceeded - could be rate limiting
                throw new translate_2.RateLimitedError(`Google Translate deadline exceeded: ${err.message}`);
            }
            throw new Error(`Google couldn't translate "${text}": ${err.message || err}`);
        }
    }
}
exports.GoogleV3Translator = GoogleV3Translator;
